ironfish/src/syncer.ts

```diff
@@ -99,46 +111,171 @@ export class Syncer {
       return
     }
 
-    if (this.state === 'idle') {
-      this.findPeer()
+    switch (this.state) {
+      case 'idle': {
+        await this.findPeer(null)
+        break
+      }
+
+      case 'measuring': {
+        await this.findPeer(this.lastLoaderIdentity)
+        if (this.loader) {
+          this.numberOfMeasurements += 1
+        }
+        break
+      }
+
+      case 'syncing': {
+        if (this.nextMeasureTime >= performance.now()) {
+          break
+        }
+
+        this.logger.info('Checking for a potentially better peer to sync from')
+
+        // If it is time to enter the measuring state, we stop syncing and set
+        // the state to measuring. This gives the syncer time to finish any
+        // in-flight requests
+        this.state = 'measuring'
+        if (this.loader) {
+          this.lastLoaderIdentity = this.loader.getIdentity()
+          await this.wait()
+        }
+        break
+      }
     }
 
-    this.eventLoopTimeout = setTimeout(() => this.eventLoop(), SYNCER_TICK_MS)
+    this.eventLoopTimeout = setTimeout(() => void this.eventLoop(), SYNCER_TICK_MS)
   }
 
-  findPeer(): void {
+  /**
+   * Chooses a peer to sync from based on measuring the connection of a random
+   * sampling of connected peers and begins syncing
+   */
+  async findPeer(currentPeerIdentity: Identity | null): Promise<void> {
     const head = this.chain.head
-
     if (!head) {
       return
     }
 
+    this.logger.debug('Syncer is beginning peer candidate measurements')
+    const measurementStart = BenchUtils.start()
+
     // Find all allowed peers that have more work than we have
     const peers = this.peerNetwork.peerManager
       .getConnectedPeers()
       .filter((peer) => peer.features?.syncing && peer.work && peer.work > head.work)
 
-    // Get a random peer with higher work. We do this to encourage
-    // peer diversity so the highest work peer isn't overwhelmed
-    // as well as helping to make sure we don't get stuck with unstable peers
-    if (peers.length > 0) {
-      const peer = ArrayUtils.sampleOrThrow(peers)
+    if (peers.length === 0) {
+      return
+    }
+
+    // If there is only one valid peer to sync from, there is no point in
+    // measuring the connection so begin syncing immediately
+    if (peers.length === 1) {
+      this.startSync(peers[0])
+      return
+    }
+
+    const syncCandidates = ArrayUtils.shuffle(peers)
+
+    // If we have been syncing from a peer, we want to include this peer in the
+    // measurement. This will allow us to maintain a connection to a strong peer
+    // if we have found one.
+    if (
+      currentPeerIdentity &&
+      !syncCandidates.find((p) => p.getIdentity() === currentPeerIdentity)
+    ) {
+      const currentPeer = this.peerNetwork.peerManager.getPeer(currentPeerIdentity)
+      if (currentPeer && currentPeer.state.type === 'CONNECTED') {
+        syncCandidates.push(currentPeer)
```

Comment:
Since we're including all connected peers with more work than us, `currentPeer` is likely to be in `syncCandidates` already. If we want it to be included in one of the 8, I think we should move it to the front of the array.

Reply:
Actually wondering if it would be included. We have [this line](https://github.com/iron-fish/ironfish/blob/staging/ironfish/src/syncer.ts#L449) where we set the peer's work when we sync a block to them. So possibly after we start syncing from someone their work will always be exactly the same as ours?

Reply:
Oh yep, we update the work if the sequence is greater than the peer's sequence. On initial connect to a peer, the peer's sequence + work will likely be HEAD, but I think there are several cases where it might lag behind what blocks the peer actually has. Probably another area we could follow up on. So two cases to handle then:

1. If the peer is in `syncCandidates`, move them to the front of the array
2. If the peer is not in `syncCandidates`, insert them at the front of the array